sys/net/link_layer/ieee802154/ieee802154_frame.c

```diff
@@ -125,85 +111,59 @@ uint8_t ieee802154_frame_init(ieee802154_frame_t *frame, uint8_t *buf)
   */
 uint8_t ieee802154_frame_get_hdr_len(ieee802154_frame_t *frame)
 {
-    uint8_t len = 0;
-
-    if (frame->fcf.dest_addr_m == 0x02) {
-        len += 2;
-    }
-    else if (frame->fcf.dest_addr_m == 0x03) {
-        len += 8;
-    }
+    uint8_t len = 3;    /* FCF and DSN have fixed length of 3 byte */
 
-    if (frame->fcf.src_addr_m == 0x02) {
-        len += 2;
+    if ((frame->fcf & IEEE_802154_FCF_DST_ADDR_MASK) == IEEE_802154_FCF_DST_ADDR_SHORT) {
+        len += 4;       /* 2 byte short address + 2 byte destination PAN id */
     }
-    else if (frame->fcf.src_addr_m == 0x03) {
-        len += 8;
+    else if ((frame->fcf & IEEE_802154_FCF_DST_ADDR_MASK) == IEEE_802154_FCF_DST_ADDR_SHORT) {
+        len += 10;      /* 8 byte long address + 2 byte destination PAN id */
     }
 
-    if ((frame->fcf.dest_addr_m == 0x02) || (frame->fcf.dest_addr_m == 0x03)) {
-        len += 2;
+    if ((frame->fcf & IEEE_802154_FCF_SRC_ADDR_MASK) == IEEE_802154_FCF_SRC_ADDR_SHORT) {
+        len += 4;       /* 2 byte short address + 2 byte source PAN id */
     }
-
-    if ((frame->fcf.src_addr_m == 0x02) || (frame->fcf.src_addr_m == 0x03)) {
-        len += 2;
+    else if ((frame->fcf & IEEE_802154_FCF_SRC_ADDR_MASK) == IEEE_802154_FCF_SRC_ADDR_LONG) {
+        len += 10;      /* 8 byte long address + 2 byte source PAN id */
     }
 
-    /* if src pan id == dest pan id set compression bit */
-    if (frame->src_pan_id == frame->dest_pan_id) {
-        frame->fcf.panid_comp = 1;
+    /* see if PAN id is compressed (src_pan_id := dst_pan_id) */
+    if (frame->fcf & IEEE_802154_FCF_PANID_COMP) {
         len -= 2;
     }
 
-    /* (DPID + DAD + SPID + SAD) + (FCF + DSN) */
-    return (len + 3);
+    return len;
 }
 
 uint8_t ieee802154_frame_read(uint8_t *buf, ieee802154_frame_t *frame,
                               uint8_t len)
```

Comment:
Did not look too deep into it, but is this function still necessary with your change? Wouldn't it suffice to do

```C
ieee802154_frame_t *frame = (ieee802154_frame_t *)buf;
```